feat: major refactor, add Python, Java, C++ support

davidli3100 opened the pull request:
Refactored searching, sorting, and recursion submodules to sort by languages

**TODO**

Python:
- [x] Recursion
- [x] Searching
- [x] Sorting

Java:
- [x] Recursion
- [x] Searching
- [x] Sorting

C++:
- [x] Recursion
- [x] Searching
- [x] Sorting

davidli3100 committed:
feat: major refactor, add python recursion

- refactored recursion, searching, sorting into language-based folders
- added python equivalents for recursion

davidli3100 committed:
docs: fix docstring indentation

davidli3100 committed:
feat (python): add linear search

davidli3100 committed:
feat: add recursion, fix misc. typos

- refactor recursion into language subfolders
- add recursion for python
- fix misc. typos in documentation

davidli3100 committed:
feat(python): bubble sort

davidli3100 committed:
feat(python): add insertion sort

davidli3100 committed:
feat (python): add selection sort

davidli3100 committed:
feat(java): add recursion

davidli3100 committed:
feat(java): add recursive binary search

davidli3100 committed:
docs:  update docstrings

davidli3100 committed:
feat (java): add non recursive binary search

davidli3100 committed:
chore: remove unused java.lang.math import

davidli3100 committed:
feat (java): add linear search

davidli3100 committed:
feat (java): add bubble sort

davidli3100 committed:
feat(java): add insertion sort

davidli3100 committed:
docs: add import comment

davidli3100 committed:
feat (java): add selection sort

davidli3100 committed:
feat (c++): add recursion

davidli3100 committed:
docs: add docstrings to c++ recursion

davidli3100 committed:
docs: fix spelling

davidli3100 commented:
Ready to merge if it looks good to you @mrseidel - I can add searching/sorting for c++ based on whether or not you want to use vectors or just plain C arrays in those.

(C arrays get a little sketchy since you have to start doing `sizeof(array)/sizeof(array[0])` to get the length of the array)

davidli3100 committed:
docs: fix malformatted comments

mrseidel commented:
> Ready to merge if it looks good to you @mrseidel - I can add searching/sorting for c++ based on whether or not you want to use vectors or just plain C arrays in those.
> 
> (C arrays get a little sketchy since you have to start doing `sizeof(array)/sizeof(array[0])` to get the length of the array)

Vectors are easier to work with overall, so go in that direction.  I'll review this pull request tomorrow.

davidli3100 committed:
feat(c++): add recursion examples

davidli3100 committed:
feat(c++): add linear search

davidli3100 committed:
feat(c++): add bubble sort

davidli3100 committed:
docs: update docstrings for bubble sort

davidli3100 committed:
fix: remove .at for vectors

- using vector::operator[] instead
- no need for boundary checking at this scale

davidli3100 committed:
feat(c++): add insertion sort

davidli3100 committed:
feat(c++): add selection sort

davidli3100 commented:
Implemented searching/sorting using vectors - I used `vector::operator[]` instead of `vector::at` just so it causes minimal confusion - not sure if boundary checking really matters here since it's just some basic algorithms

mrseidel reviewed (changes requested):
Let's use this pull request to also refactor the other folders (since you're completing major refactoring anyways) to ensure a single style (i.e. instead of c-plus-plus.cpp change into /c++/main.cpp)

- [ ] If you could rename the other files/folders in previous notes to match your standard.  For C++ and C#, I know that filenames cannot have those symbols, but folders can.

davidli3100 committed:
chore: refactor folder names to match new changes

davidli3100 commented:
Refactored folder names to use `C++` and `C#` - I left the file names the same since they match the class names
